src/Network/Transport/TCP.hs

```diff
@@ -1032,7 +1034,7 @@ handleConnectionRequest transport socketClosed (sock, sockAddr) = handle handleE
         -- been recorded as part of the remote endpoint. Either way, we no longer
         -- have to worry about closing the socket on receiving an asynchronous
         -- exception from this point forward.
-        forM_ mEndPoint $ handleIncomingMessages (transportParams transport) . (,) ourEndPoint
+        return $ fmap (handleIncomingMessages (transportParams transport) . (,) ourEndPoint) mEndPoint
```

Comment:
Good catch. This problem lurks in master as well, probably since connection timeouts were introduced. Fix coming shortly.